content/frontend/ember-apollo/5-authentication.md

```diff
@@ -0,0 +1,691 @@
+
+```bash
+ember generate service auth
+```
+
+The `userId` and `authToken` will be stored in the user’s `localStorage` in this app, so add some methods for setting and retrieving that data from `localStorage` and in your service:
+
+```js(path=".../hackernews-ember-apollo/app/services/auth.js")
+import Ember from 'ember';
+
+// 1.
+const GC_USER_ID = 'graphcool-user-id';
+const GC_AUTH_TOKEN = 'graphcool-auth-token';
+
+export default Ember.Service.extend({
+  init() {
+    this._super(...arguments);
+    // 2.
+    this.getUserId();
+    this.getAuthToken();
+  },
+
+  authToken: null,
+
+  // 3.
+  getUserId() {
+    const userId = localStorage.getItem(GC_USER_ID);
+    this.setUserId(userId);
+    return userId;
+  },
+
+  getAuthToken() {
+    const token = localStorage.getItem(GC_AUTH_TOKEN);
+    this.setAuthToken(token);
+    return token;
+  },
+
+  // 4.
+  removeUserId() {
+    localStorage.removeItem(GC_USER_ID);
+    this.set('userId', null);
+  },
+
+  removeAuthToken() {
+    localStorage.removeItem(GC_AUTH_TOKEN);
+    this.set('authToken', null);
+  },
+
+  // 5.
+  setUserId(id) {
+    localStorage.setItem(GC_USER_ID, id);
+    this.set('userId', id);
+  },
+
+  setAuthToken(token) {
+    localStorage.setItem(GC_AUTH_TOKEN, token);
+    this.set('authToken', token);
+  },
+
+  userId: null
+});
+```
+
+</Instruction>
+
+Let’s breeze through what is happening in this piece of code as it’s vital to the application in the long run.
+
+1. You are defining two constants, `GC_USER_ID` and `GC_AUTH_TOKEN` which hold the key values that `localStorage` will need.
+2. Right off the bat in the service initialization, you are executing two methods to get the user’s ID and auth token, if they are set in `localStorage`.
+3. You are also defining those two methods `getUserId` and `getAuthToken` which query the `localStorage` for the value and pass it to each method to be set on your service.
+4. Two methods, `removeUserId` and `removeAuthToken`, are created to remove the value from `localStorage` and reset the local value, on the service, to null.
+5. Finally, you have the two methods that are setting the value in `localStorage` and on the service.
+
+What’s missing from this long block of code is methods to login, signup, and logout. You will add those in a moment, but first let’s talk about how Graphcool handles authentication.
+
+### Enabling Email-and-Password Authentication & Updating the Schema
+
+<Instruction>
+
+In the directory where `project.graphcool` is located, type the following into the terminal:
+
+```bash
+graphcool console
+```
+
+</Instruction>
+
+This will open up the Graphcool Console - the web UI that allows you to configure your Graphcool project.
+
+<Instruction>
+
+Select the *Integrations*-tab in the left side-menu and then click on the *Email-Password-Auth*-integration.
+
+</Instruction>
+
+![](http://imgur.com/25iMZtf)
```

Comment:
the image here is not rendered, you need to refer to it by adding `png` as a suffix: `http://i.imgur.com/25iMZtf.png`